ReadableStream supports for-await-of: minimal doc (#24310)
mdn/content: files/en-us/web/api/readablestream/index.md

## Async iteration

`ReadableStream` implements the [async iterable protocol](/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols).
This enables asynchronous iteration over the chunks in a stream using the [`for await...of`](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) syntax:

```js
const stream = new ReadableStream(getSomeSource());

for await (const chunk of stream) {
  // Do something with each 'chunk'
}
```

The async iterator consumes the stream until it runs out of data or otherwise terminates.
The loop can also exit early due to a `break`, `throw`, or `return` statement.

While iterating, the stream is locked to prevent other consumers from acquiring a reader (attempting to iterate over a stream that is already locked will throw a `TypeError`).
This lock is released when the loop exits.

By default, exiting the loop will also cancel the stream, so that it can no longer be used.
To continue to use a stream after exiting the loop, pass `{ preventCancel: true }` to the stream's `values()` method:

```js
for await (const chunk of stream.values({ preventCancel: true })) {
  // Do something with 'chunk'

>   return;

Reviewer: ```suggestion
  break;
```

- {{domxref("ReadableStream.tee()")}}
  - : The `tee` method [tees](https://streams.spec.whatwg.org/#tee-a-readable-stream) this readable stream, returning a two-element array containing the two resulting branches as new {{domxref("ReadableStream")}} instances. Each of those streams receives the same incoming data.

> ## Async Iteration

Reviewer: ```suggestion
## Async iteration
```

Author: Duh. Thank you.

> `ReadableStream` supports asynchronous iteration over the chunks in a stream using the [for await...of](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) syntax:
> 

Author: ```suggestion
`ReadableStream` implements the [async iterable protocol](/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols).
This enables asynchronous iteration over the chunks in a stream using the [for await...of](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) syntax:

```

> This enables asynchronous iteration over the chunks in a stream using the [for await...of](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) syntax:

Reviewer: ```suggestion
This enables asynchronous iteration over the chunks in a stream using the [`for await...of`](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) syntax:
```

```js
const stream = new ReadableStream(getSomeSource());

for await (const chunk of stream) {
  // Do something with each 'chunk'
}
```

The async iterator consumes the stream until it runs out of data or otherwise terminates.

> The loop can also exit early due to a `break`, `throw`, or `return`.

Reviewer: ```suggestion
The loop can also exit early due to a `break`, `throw`, or `return` statement.
```

While iterating, the stream is locked to prevent other consumers from acquiring a reader (attempting to iterate over a stream that is already locked will throw a `TypeError`).

> This lock is released when the iterator exits the loop.

Reviewer: ```suggestion
This lock is released when the loop exits.
```

By default, exiting the loop will also cancel the stream, so that it can no longer be used.
To continue to use a stream after exiting the loop, pass `{ preventCancel: true }` to the stream's `values()` method:

```js
for await (const chunk of stream.values({ preventCancel: true })) {
  // Do something with 'chunk'
  return;
}
// Acquire a reader for the stream and continue reading ...
```

Notes:

> - `ReadableStream` implements the [async iterable protocol](/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols).

Reviewer: This should be moved to the opening of the section

Author: I moved it here deliberately - as a user I probably don't care all that much while I know that I can do the for loop.

Reviewer: It's because the "async iterable protocol" has more implications than `for await...of` loops. For example, with the [async-iterator-helpers](https://github.com/tc39/proposal-async-iterator-helpers) proposal, you will also be able to use async iterator methods on the returned iterator.

Author: But Let's try that.

Author: THanks for the explanation. MOved.

> - `ReadableStream` implements the [async iterable protocol](/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols).

Author: ```suggestion
```

> - You can abort processing of a stream prematurely triggering code to call `throw` or `break` in the loop.

Reviewer: I guess you mean "you can abort processing of a stream prematurely by using `throw` or `break` in the loop." In this case, this seems duplicated from the sentence above "The loop can also exit early due to a `break`, `throw`, or `return`." I guess you can merge this note item with the above sentence.

Author: This "goes with" the next sentence. 

What it is saying is that you might want to have a trigger (such as a button press) that you signal in the body of the for loop and then call return. But this will not be instant - the loop code is only run when you get more data, and that is indeterminate.

So yes, it is text that was said before, but it belongs here to make the point. We could remove the point if you think it is ... pointless :-)

Author: How about https://github.com/mdn/content/pull/24310/files#r1102333925

>   If code from outside the loop triggers code in the loop, such as a `break`, there may be a delay before the code in the loop is executed.

Reviewer: I don't quite get this point. An example may help

Author: I'm going to remove this, and I'll put a proper example in the "Using readable streams" in the follow up PR.

But FYI, if you click the button to signal the abortcontroller, this is then will be acted on inside the loop. If your next chunk took 10 seconds to arrive, then the abort would happen in 10 seconds.
```
let bytes = 0;
logChunks(stream);

const aborter = new AbortController();
button.addEventListener("click", () => aborter.abort());
logChunks(stream, { signal: aborter.signal });

async function logChunks(readableXX, { signal }) {
  try {
    for await (const chunk of readableXX) {
      if (signal.aborted) throw signal.reason;
      bytes += chunk.length;
      logConsumer(`Chunk: ${chunk}. Read ${bytes} characters.`);
    }
  } catch (e) {
    if (e instanceof TypeError) {
      logConsumer("TypeError: Browser may not support async iteration");
    } else {
      logConsumer(`Error in async iterator: ${e}.`);
    }
  }
}
```
It is possible to cancel your stream faster by aborting the underlying source rather than using this kind of approach. However not every source is abortable, and apparently if you have streams feeding into streams you want to cancel in the final consumer, not the start of the stream.

> - You can abort processing of a stream prematurely triggering code to call `throw` or `break` in the loop.
>   Note however that as each iteration is only run when the next chunk arrives, there can be a delay between signalling that you want to cancel, and the code getting to run.
>   The fastest way to terminate the stream may be to abort the "underlying source" of the stream.

Author: @Josh-Cena How about this then?

```suggestion
- The code in the loop is only run when the next chunk arrives.
  If a user signal some operation from outside the loop that is implemented in the loop, such as a `break`, there may be a delay.
```

Author: @Josh-Cena ping ^^^ . In case you missed it - if you're just too busy this can wait until you are ready.

Reviewer: Sorry, I'm quite busy this weekend. I'll try to find time next Monday, or before if I can.

Author: No worries at all. Have a good one.

> - Polyfills are available if this feature is not [supported on your browser](#browser_compatibility): [web-streams-polyfill](https://github.com/MattiasBuelens/web-streams-polyfill) or [sd-streams](https://github.com/stardazed/sd-streams).

Reviewer: I don't think we usually add polyfill links in the prose. Maybe move this to "See also"?

Author: Correct. I'll move it even though I do prefer it close to where you're using the feature.

This works with both async and non-async iterators.

> ### Async iteration of stream (for-wait-of)

Reviewer: ```suggestion
### Async iteration of a stream using for await...of
```

> This example shows how you can process the `fetch()` response using a [for await...of](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loop to iterate through the arriving chunks.

Author: A much better example would be to rewrite the first one in this doc using the new form.
My concern though is that I might screw it up, because I don't know how to test a fetch.

IN particular, reading would be easy; it's the formatting to output a new stream I'm not sure of.

I'm happy to look at that as a post process, but would be good to get this in as it provides an end point link for release notes and BCD mdn URL

> This example shows how you can process the `fetch()` response using a [for await...of](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loop to iterate through the arriving chunks.

Reviewer: ```suggestion
This example shows how you can process the `fetch()` response using a [`for await...of`](/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loop to iterate through the arriving chunks.
```
